M17 GNSS Meta

turnrye opened the pull request:
# What

This change adds support for sending M17 GNSS meta.

# How

This change is based on the work originally done in https://github.com/OpenRTX/OpenRTX/pull/314, current as of https://github.com/OpenRTX/OpenRTX/pull/314/commits/e8c3ea54b93fe894b86d69c29448fdcdb236bb7d. The changes there had whitespace changes removed, then were squash merged onto a new branch. From there, changes that were unrelated to the M17 GNSS meta feature were removed. This yielded the changes in this PR.

The work in this PR was done by @kd0oss Rick. I am helping pick and land the changes.

I've left this in an un-squashed state intentionally so that both the reviewers and Rick can "follow along" with how this branch was created. This PR is trying to avoid rebasing since there are multiple contributors. It should be squashed _after review_ but before merging. This strategy will be followed again for picking additional changes from this fork into the mainline repo.

kd0oss commented on openrtx/src/rtx/OpMode_M17.cpp:
```diff
@@ -172,7 +179,9 @@ void OpMode_M17::offState(rtxStatus_t *const status)
 
 void OpMode_M17::rxState(rtxStatus_t *const status)
 {
```
That seems reasonable.

silseva commented on openrtx/include/rtx/OpMode_M17.hpp:
```diff
@@ -134,20 +143,34 @@ class OpMode_M17 : public OpMode
+    /**
+     * Convert OpenRTX latitude/longitude format into M17 format and vice versa.
+     */
+    void rtx_to_q(int32_t* qlat, int32_t* qlon, int32_t lat, int32_t lon);
```
These are actually general-purpose conversion functions: please move them to gps.h or utils.h
Also, there is no "OpenRTX latitude/longitude" and "M17" formats. OpenRTX is using fixed point, while M17 is using Q1.24

silseva commented on openrtx/src/protocols/M17/M17LinkSetupFrame.cpp:
```diff
@@ -76,6 +76,12 @@ void M17LinkSetupFrame::setType(streamType_t type)
     data.type  = type;
 }
 
+// Added by KD0OSS (2025)
```
This is adding meta text, but this PR deals with GNSS data. I'd remove these changes and add them to the other incoming PRs

turnrye replied:
Thanks, I can move this out!

silseva commented on openrtx/src/rtx/OpMode_M17.cpp:
```diff
@@ -16,21 +16,27 @@
 
 #ifdef PLATFORM_MOD17
 #include <calibInfo_Mod17.h>
-#include <interfaces/platform.h>
```
Restore this line

silseva commented on openrtx/src/rtx/OpMode_M17.cpp:
```diff
@@ -335,6 +347,93 @@ void OpMode_M17::txState(rtxStatus_t *const status)
+        gpsTimer = 0;
+
+        gps_t gps_data;
+        pthread_mutex_lock(&state_mutex);
```
Locking the state mutex in the rtx threads may introduce delays disrupting the correct timings for M17 modulation. GPS data is being fetched for read access and is actually safe to retrieve it without a critical section

silseva commented on openrtx/src/rtx/OpMode_M17.cpp:
```diff
@@ -335,6 +347,93 @@ void OpMode_M17::txState(rtxStatus_t *const status)
+            platform_ledOn(YELLOW); // Blink LED yellow when sending GNSS
+        	uint8_t gnss[14] = {0};
+
+        	gnss[0] = (M17_GNSS_SOURCE_OPENRTX<<4) | M17_GNSS_STATION_HANDHELD; //OpenRTX source, portable station
```
This code should be moved into a dedicated member function of the M17LinkSetupFrame class. For example `void setGnssData(const struct gps_t *position)`

turnrye commented on openrtx/include/protocols/M17/M17Demodulator.hpp:
```diff
@@ -165,7 +165,9 @@ class M17Demodulator
     filter_state_t                 dcrState;        ///< State of the DC removal filter
 
     Correlator   < M17_SYNCWORD_SYMBOLS, SAMPLES_PER_SYMBOL > correlator;
+    Synchronizer < M17_SYNCWORD_SYMBOLS, SAMPLES_PER_SYMBOL > lsfSync   {{ +3, +3, +3, +3, -3, -3, +3, -3 }};
```
Thanks, I can move this out! This isn't needed till the PR for packet mode for SMS.

turnrye commented on openrtx/src/protocols/M17/M17Demodulator.cpp:
```diff
@@ -265,17 +265,24 @@ bool M17Demodulator::update(const bool invertPhase)
                 case DemodState::UNLOCKED:
                 {
                     int32_t syncThresh = static_cast< int32_t >(corrThreshold * 33.0f);
-                    int8_t  syncStatus = streamSync.update(correlator, syncThresh, -syncThresh);
```
Thanks, I can move this out! This isn't needed till the PR for packet mode for SMS.

silseva commented on openrtx/src/protocols/M17/M17LinkSetupFrame.cpp:
```diff
@@ -158,3 +159,55 @@ uint16_t M17LinkSetupFrame::crc16(const void *data, const size_t len) const
+
+void M17LinkSetupFrame::setGnssData(const struct gps_t *position) {
+    // TODO: shift this to a struct with bitwise fields
+    uint8_t gnss[14] = {0};
```
Yes, is fine. When you do this on a type which is not a basic one, the compiler is replacing that line with a `memset`.

turnrye commented on openrtx/include/protocols/M17/M17Datatypes.hpp:
```diff
@@ -28,6 +28,13 @@
 #error This header is C++ only!
 #endif
 
+// Constants for the GNSS Meta field
```
Can't use the enum within a bit field though, so they have to stay this way

silseva replied:
You can :)
See here: https://github.com/OpenRTX/OpenRTX/blob/4b9f75fe81092947eaef586d304cf10662cbb6ff/openrtx/include/core/settings.h#L57
`display_timer` is a bitfield but it its assigned a value from an `enum`

turnrye replied:
Thanks! Forgot that I didnt have to set the enum type on the bitfield itself. Eventually since we want to use these in a target hwconfig file, which is C, I think I need to move these enums out of this file. Where would they belong best?

silseva replied:
Let's leave the things as they're for now, is already good :)

silseva commented on openrtx/src/rtx/OpMode_M17.cpp:
```diff
@@ -335,6 +354,47 @@ void OpMode_M17::txState(rtxStatus_t *const status)
         modulator.sendFrame(m17Frame);
     }
 
+    if(lsfFragCount == 6)
```
Thanks. Actually this logic can be moved inside the FrameEncoder, to have it set up once for all the code touching the LSF.

turnrye commented:
I'll work on squashing and rebasing this, since at this point the implementation is relatively stable and @silseva plans to take up the final steps.

Separate note, a few of us have noticed that at times the GPS state is set as having a fix yet the lat and lon are incorrectly 0. There may be a bug around this, perhaps outside of this PR, that needs to be addressed.

turnrye commented:
> > When we add the meta text option, I think we should also add a settings entry for controlling what meta is sent (none, text, gnss, text+gnss). But since GNSS already has a dedicated menu option, I'd think this is OK to go to master for now, and we just make sure to add that change before the feature is in a release.
> 
> Can't you get that by setting meta text (or not) and turning the GPS on (or not)? What's the use case for setting meta text and then not transmitting it? I guess you could turn on the GPS to see your lat/long or set the clock, while not wanting it to transmit, but that seems like a pretty marginal case.

Yes, you're right that it can be managed through those other settings. Let's discuss more on the next PR once that's up (~ end of this week hopefully).

turnrye commented:
I've connected with @silseva, and we agreed for me to give the LSF automated chunk update changes a try. I'll be working on those next and continuing to rewrite history a bit as the sole developer on this branch.

turnrye commented:
Hey @silseva this is ready for review again!

Re identifying when it's safe to update LSF, I kept with the simple flag based approach rather than adding a second buffer to the encoder. I felt this performed well enough in my testing. I also recognize we have the chance to revisit this with my next PR for text metadata.

silseva commented on openrtx/src/rtx/OpMode_M17.cpp:
```diff
@@ -335,6 +349,28 @@ void OpMode_M17::txState(rtxStatus_t *const status)
 
+    gps_t gps_data;
+    gps_data = state.gps_data;
+    if(encoder.canUpdateLsf() && state.settings.gps_enabled && (gpsTimer == -1 || gpsTimer >= MINIMUM_GPS_TIME_TICKS) && gps_data.fix_type > FIX_TYPE_NOT_AVAIL)
```
Two things:

1. For the first initialization, I'd just put 
```
if(state.settings.gps_enabled)
    lsf.setGnssData(&gps_data, M17_GNSS_STATION_HANDHELD);
```
in the block preparing the LSF for the first transmission. This saves you from having to set `gpsTimer` to -1 and have a cumbersome condition checking.

2. Move the `gps_data.fix_type > FIX_TYPE_NOT_AVAIL` check inside the `setGnssData` function: there is the place where you're composing the data block and setting the various validity flags

silseva commented on openrtx/src/rtx/OpMode_M17.cpp:
```diff
@@ -335,6 +349,28 @@ void OpMode_M17::txState(rtxStatus_t *const status)
+        gpsTimer = 0;
+        gpsTransmitting = true;
+        lsf.setGnssData(&gps_data, M17_GNSS_STATION_HANDHELD);
+        type.fields.encType    = M17_ENCRYPTION_NONE;
```
I'd move the setting of the LSF type field inside the `setGnssData` function itself: being it a member function of the M17LinkSetupFrame class, it has access to all the fields of the LSF data structure.

silseva commented on openrtx/src/protocols/M17/M17LinkSetupFrame.cpp:
```diff
@@ -158,3 +159,40 @@ uint16_t M17LinkSetupFrame::crc16(const void *data, const size_t len) const
+    uint16_t alt = (uint16_t)1000 + position->altitude * 2;
+    gnssData.altitude = __builtin_bswap16(alt);
+
+    data.meta.gnss_data = gnssData;
```
Instead of creating a local copy of  `gnssData` on the stack, filling it data and then _copying_ it into the corresponding field of the LSF, I'd just access `data.meta.gnss_data` directly.

silseva commented on openrtx/include/rtx/OpMode_M17.hpp:
```diff
@@ -148,6 +149,10 @@ class OpMode_M17 : public OpMode
+    M17::M17LinkSetupFrame lsf;        ///< M17 link setup frame
+
+
+    static constexpr int16_t MINIMUM_GPS_TIME_TICKS = 5 * 30; // Wait at least 5 seconds between GPS transmissions
```
Do you mind moving this above all the variables' declarations? Usually I tend to have:
- private member functions
- private constants
- private variables

turnrye commented:
@silseva thanks for the quick feedback! I've addressed the issues you raised and left it as a separate fixup commit so that you can trace the individual changes if you wish. I've also verified that the changes still work with a quick "happy path" test. Please take another look at your convenience and let me know if there is any further feedback.
